src/job.rs

```diff
@@ -1,16 +1,102 @@
+use anyhow::{bail, Context, Result};
 use std::collections::HashMap;
-use std::io;
+use std::fs;
+use std::path::{Path, PathBuf};
 use std::process::{Command, Output};
+use tempfile::Builder;
+use walkdir::WalkDir;
 
 #[derive(Debug, PartialEq, Eq)]
 pub struct Job {
     pub command: String,
     pub arguments: Vec<String>,
     pub environment: HashMap<String, String>,
+    pub working_directory: PathBuf,
+    pub inputs: Vec<PathBuf>,
+    pub outputs: Vec<PathBuf>,
 }
 
 impl Job {
-    pub fn run(&self) -> io::Result<Output> {
+    pub fn run(&self) -> Result<Output> {
+        let work_dir = Builder::new()
+            .prefix(format!("job-{}", self.command).as_str())
+            .tempdir()
+            .context("couldn't create a temporary directory for the job")?;
+
+        self.prepare_workspace(work_dir.path())
+            .context("couldn't prepare files in the working directory")?;
+
+        let output = self
+            .run_job(work_dir.path())
+            .context("couldn't run the job")?;
+
+        self.tear_down_workspace(work_dir.path())
+            .context("couldn't tear down the workspace")?;
+
+        work_dir
+            .close()
+            .context("couldn't clean up the job's temporary directory")?;
+
+        Ok(output)
+    }
+
+    fn prepare_workspace(&self, work_dir: &Path) -> Result<()> {
+        for input in &self.inputs {
+            let meta = fs::metadata(input)
+                .with_context(|| format!("couldn't read metadata for {}", input.display()))?;
```

Comment:
🎨 You should be able to leave off the call to `.display()` here (and elsewhere in the file) - the `format!` macro should call `display` for you because whenever it sees `{}` in the string literal, it expects that what you give it will have the [`Display`](https://doc.rust-lang.org/std/path/struct.Display.html) trait and therefore a `.display()` method.